numpy/core/src/multiarray/arrayobject.c

```diff
@@ -511,7 +507,9 @@ array_dealloc(PyArrayObject *self)
             }
             // Guess at malloc/free ???
             free(fa->data);
-        } else {
+        }
+        else {
+            size_t nbytes = PyArray_NBYTES_ALLOCATED(self);
```

Comment:
I guess this can't work unless all places that allocate do it right, and I am not sure how many there are (too small or 0 is fine, the cache can deal with it).
Maybe the only thing left for now is to `free` using `NBYTES` (also in that weird `setstate`), but use `NBYTES_ALLOCATED` for allocating in `setstate` to fix the ref-leak.

Or... ignore this one for now, because leaking the memory handler on unpickling doesn't matter anyway in practice, its a bug, but just not a huge deal.

Reply:
> unless all places that allocate do it right

Places that don't do it right should be seeing a "uhoh", and so far the only case that emits "uhoh" is for 0 in strides. On the other hand, this PR is segfaulting, so maybe something is off.

Reply:
Ah OK, so the actual problematic cases are fewer than I thought I guess.  Although, now I am a bit worried that we might create "too small" arrays somewhere, but don't cover it in the tests.

Reply:
I guess one possibility is that we are hitting some bad path only in the full test suite.